### ISSUE TITLE
Slow performance

### ISSUE THREAD

**Issue by pensierinmusica** [brianc/node-postgres#861]
Hi,

I'm using `node-postgres` (v4.4.1), with Node (v4.1.0), and postgres (v9.4.4) on a MacBook Pro (2.3 GHz Intel Core i7 processor, 16 GB 1600 MHz DDR3 memory, and 500 GB Flash Storage).

Benchmarking it against [`pgbench`](http://www.postgresql.org/docs/9.4/static/pgbench.html), there seems to be a huge performance difference.

(I haven't tested using `pg-native`, since it's not supported with Node >= v4 yet, but according to your docs it can maximum bring a 20-30% increase in parsing speed).

**Pgbench is ~40k tps** whereas **node-postgres is ~5k tps** (in both cases performing the same simple selection, with the same number of clients).

This is the pgbench code:
```
~ $ pgbench test -c 5 -t 10000 -S
starting vacuum...end.
transaction type: SELECT only
scaling factor: 10
query mode: simple
number of clients: 5
number of threads: 1
number of transactions per client: 10000
number of transactions actually processed: 50000/50000
latency average: 0.000 ms
tps = 41367.236818 (including connections establishing)
tps = 42233.261058 (excluding connections establishing)
```

And this is the node-postgres code:
```js
var pg = require('pg');
var conString = "postgres://admin:admin@localhost/test";

pg.defaults.poolSize = 5;

pg.connect(conString, function(err, client, done) {

  var start = new Date().getTime();
  var transactions = 50000;
  var count = 0;

  for (var i = 0; i < transactions; i++) {
    client.query('SELECT * FROM "Users" WHERE id = \'2f34d8ca-9636-406c-b6e0-2ea2b0c39061\'', function(err, result) {
      count++;
      if (count === transactions) {
        var end = new Date().getTime();
        console.log('Elapsed ms: ' + (end - start));
        console.log(result.rows[0]);
        done();
      }
    });
  }

});
```

Is this normal / expected? Why is there such a huge difference between the performance of the two (8x factor) and what could be done to reduce it?

Thanks!

**Comment by PritamUpadhyay** [comment]
+1

**Comment by brianc** [comment]
Without seeing your code & knowing exactly what pgbench is doing I wouldn't
be able to tell you.  Node-postgres itself is pretty lightweight and 'low
level' introducing the minimum amount of overhead.  Likely some of the
speed difference is related to node itself, the JIT not having time to warm
up, the way your system is configured, GC overhead in node, etc etc.
Benchmarking is a tricky thing and throwing together one and then opening
up an issue doesn't really do much good.  There are a __lot__ of different
things that can affect speed.

If you think there is significant slowdown somewhere in node-postgres I
recommend you dive in, do some performance testing, benchmarking,
profiling, and see if you can find any hot code paths that can be
optimized.  Once you find some, I'll be happy to review any pull requests
related to significant speed improvements. :thumbsup:

**Comment by pensierinmusica** [comment]
Hi @brianc, thanks for the quick reply. I didn't mean to criticize any of node-postgres, but simply trying to figure out where the problem is. As for looking into the codebase myself from scratch, that would require a big amount of time before I even get close to a fraction of your understanding/knowledge of it, and it's off scope for me at the moment.

Regarding pgbench, looking at the docs from the link in my original post:
>What is the "Transaction" Actually Performed in pgbench?

>The default transaction script issues seven commands per transaction:

>BEGIN;

>UPDATE pgbench_accounts SET abalance = abalance + :delta WHERE aid = :aid;

>SELECT abalance FROM pgbench_accounts WHERE aid = :aid;

>UPDATE pgbench_tellers SET tbalance = tbalance + :delta WHERE tid = :tid;

>UPDATE pgbench_branches SET bbalance = bbalance + :delta WHERE bid = :bid;

>INSERT INTO pgbench_history (tid, bid, aid, delta, mtime) VALUES (:tid, :bid, :aid, :delta, CURRENT_TIMESTAMP);

>END;

>If you specify -N, steps 4 and 5 aren't included in the transaction. If you specify -S, only the SELECT is issued.

Since we're using the `-S` option above it simply performs:
```sql
SELECT abalance FROM pgbench_accounts WHERE aid = :aid;
```

Which is basically the same as my NodeJS code is doing.

I hoped you could shed any light on this...   :)

**Comment by sehrope** [comment]
@pensierinmusica Your code is using the same `client` object repeatedly from the same connection. That means it's using a single connection to execute 50K queries vs the pgbench code that is using 5 connections, at the same time, to issue 10K queries. That probably explains most of the performance difference. For a better comparison you should be running things in parallel with something like [`async.parallelLimit(...)`](https://www.npmjs.com/package/async#parallel).

**Comment by pensierinmusica** [comment]
Thanks for the hint @sehrope!

That explains a good chunk of the performance difference. On the other hand, even reducing pgbench to 1 client, it's still twice as fast as node-postgres with **~10k tps**.

Is there a way to bring the two to comparable performance?

```
~ $ pgbench test -c 1 -t 50000 -S
starting vacuum...end.
transaction type: SELECT only
scaling factor: 10
query mode: simple
number of clients: 1
number of threads: 1
number of transactions per client: 50000
number of transactions actually processed: 50000/50000
latency average: 0.000 ms
tps = 10244.022869 (including connections establishing)
tps = 10256.353452 (excluding connections establishing)
```

**Comment by sehrope** [comment]
The PostgreSQL wire protocol allows sending a series of queries without waiting for their results. I'm not intimately familiar with the internals of pgbench but it may be doing some type of pipelining. If you're running queries sequentially using node-postgres then it won't do that. It'd be sending the query, waiting for it to complete, fetching the results, then sending the next query. I'm sure there's other differences as well. Without going through all the code for pgbench, it'll be pretty hard to have an apples to apples comparison of what it's testing and a node-postgres perf test.

All that aside, I haven't run into any performance issues with node-postgres. If you run into an issue where your use of the driver is the bottleneck (*and not the queries you're actually executing on the server*), then you're probably doing something wrong.

**Comment by pensierinmusica** [comment]
HI @sehrope, thanks for the contribution! In my understanding the code example in my first post (https://github.com/brianc/node-postgres/issues/861#issue-107323979) is not sending queries sequentially. Instead it's firing all the requests and then receiving the responses asynchronously through the respectively registered callbacks. Am I missing something?

As for the second part of your comment:
>If you run into an issue where your use of the driver is the bottleneck (and not the queries you're actually executing on the server), then you're probably doing something wrong.

Mmm... well that's exactly the case at hand, isn't it?

Looking forward to hear your thoughts. Thanks!

**Comment by pensierinmusica** [comment]
@brianc and everyone in this thread, following up on the topic, so far I can pinpoint 3 issues.

1. Why `[pgbench](http://www.postgresql.org/docs/9.4/static/pgbench.html)` with 1 client is twice as fast as `node-postgres`? (check details here: https://github.com/brianc/node-postgres/issues/861#issuecomment-141672689).
2. Given a pool of `x` clients, could `node-postgres` parallelize DB requests using the clients that are available from the pool? In my understanding, this would achieve similar results to using `pgbench` with `x` clients (e.g. with 5 clients `pgbench` is 4x faster than with 1 client, as shown here: https://github.com/brianc/node-postgres/issues/861#issue-107323979).
3. Within a single client, would it make sense to use pipelining? Form `node-redis` docs:

>The performance of node_redis improves dramatically with pipelining, which happens automatically in most normal programs.

Looking forward to hear your toughts  :)  Cheers!

**Comment by brianc** [comment]
1.

> Likely some of the
> speed difference is related to node itself, the JIT not having time to warm
> up, the way your system is configured, GC overhead in node, etc etc.
> Benchmarking is a tricky thing and throwing together one and then opening
> up an issue doesn't really do much good.  There are a __lot__ of different
> things that can affect speed.

2. You have to actually _use_ the pool, it doesn't 'automatically' fan queries across clients as this would break transaction isolation but there are lots of docs on this & is __extremely__ common way to use the driver.  [this repo](https://github.com/brianc/node-pg-query) in fact does just that.  So yeah definitely you can tweak the number of clients you have and get parallel queries as long as results from one query don't depend on results from another - if that's the case then they'd definitely need to run one after another. :smile: 

3.  The node-postgres protocol doesn't really support pipelining.  Redis doesn't do the same kind of transaction isolation and ACID guarantees as postgres.  A single connected postgres client can have a maximum of __one__ executing command at a time - if you send a second query before the first is finished executing postgres will actually return an error because its an invalid action.

I appreciate your concern on this, but this project is pretty old and stable and is used by lots of people in production.  I've had similar conversations to this one about once a year and generally the outcome is the same.  I encourage you to dive in if you think there are wins to have in performance, and like I said, we're always interested in contributions, but I feel like discussing this more is just yak shaving a hypothetical that doesn't actually exist in production apps.

**Comment by apapacy** [comment]
I start my script with 1 000 000 inserts in a loop on the desktop .
And was surprised, that the new reconnection in the each iteration runs 10 times faster than single connection before enter the cycle and usage connection pool (various size of pool 1, 2, 10, 100, 1000, 1000000) ( 1000/100 ) .
